File: CSC 435/Homework/Assignment 5/PingClient.py
import sys, time, math
from socket import *

# Get the server hostname and port as command line arguments                  
host = sys.argv[1]
port = sys.argv[2]
timeout = 1 # in second, change as needed

# Create UDP client socket
# Note the use of SOCK_DGRAM for UDP datagram packet
clientsocket = socket(AF_INET, SOCK_DGRAM)
# Set socket timeout as 1 second
clientsocket.settimeout(timeout)
# Command line argument is a string, change the port into integer
port = int(port)  
# Sequence number of the ping message
ptime = 0  

receiveCount = 0;
times = []

# Ping for 10 times
while ptime < 10: 
	ptime += 1

	# Format the message to be sent. 
        # use time.asctime() for currTime
	currTime = str(time.asctime())

	data = "PING " + str(ptime) + " " + currTime + "\r\n"
     
	try:
		print("PING " + str(ptime) + " sent to " + host + " at " + currTime)

		# Sent time. from time.time()
		RTTb = time.time()

		# Send the UDP packet with the ping message
		clientsocket.sendto(data.encode(),(host, port))

		# Receive the server response
		message, address = clientsocket.recvfrom(1024)  
		receiveCount = receiveCount + 1

		# Received time. use time.time()
		RTTa = time.time()

		# Display the server response as an output
		print("PING received from " + host + ": seq#" + str(ptime) + " time=" + time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(RTTb)))

		# Round trip time is the difference between sent and received time
		elapsedTime = (RTTa - RTTb) * 1000
		times.append(elapsedTime)

	except:
		# Server does not response
		# Assume the packet is lost
		print ("Request timed out.")
		continue

# Gather and print statistics
dropCount = 10 - receiveCount
print("--- PING statistics ---")
print("10 packets transmitted, " + str(receiveCount) + " received, " + str((100 * dropCount / 10)) + "% packet loss")

if len(times) > 0:
	min = times[0]
	max = times[0]
	sum = 0
	
	for time in times:
		sum = sum + time
		if (time < min):
			min = time
		if (time > max):
			max = time
	
	avg = sum / receiveCount
	print("rtt min/avg/max = " + str(min) + " " + str(avg) + " " + str(max) + " ms")

else:
	print("Statistics unavailable -- no packets received.")

# Close the client socket
clientsocket.close()

File: CSC 435/Homework/Assignment 5/PingClient.java
import java.net.*;
import java.time.Instant;
import java.util.LinkedList;

public class PingClient {

	public static void main(String[] args) throws Exception {
		
		// Input validation
		if (args.length < 2) throw new Exception("Please use: PingClient host port");
		
		String host = args[0];
		int port = Integer.parseInt(args[1]);
		
		DatagramSocket clientSocket;
		DatagramPacket sendPacket;
		DatagramPacket receivePacket;
		byte[] receiveData = new byte[1024];
		byte[] sendData = new byte[1024];
		InetAddress IPAddress;
		
		// Create client socket to destination
		clientSocket = new DatagramSocket();
		clientSocket.setSoTimeout(1000); 
		IPAddress = InetAddress.getByName (host);
		
		// Initialize variables for statistics
		int totalSent = 10;
		int receiveCount = 0;
		LinkedList<Long> times = new LinkedList<Long>();
		
		for (int seq = 0; seq < totalSent; seq++) {
			
			// Format PING message
			String currentTime = Instant.now().toString();
			String sendString = "PING " + seq  + " " + currentTime + ".\r\n";
			sendData = sendString.getBytes();
			
			// Create packet and send to server
			sendPacket = new DatagramPacket(sendData, 
					                        sendData.length,  
							                IPAddress, 
							                port);
			long sendTime = System.nanoTime();
			clientSocket.send(sendPacket);
			System.out.println("PING " + seq + " sent to " + host + " at " + currentTime + ".");
			
			// Create receiving packet and receive from server	
			try {
				receivePacket = new DatagramPacket(receiveData, receiveData.length); 	
				clientSocket.receive(receivePacket);
				long receiveTime = System.nanoTime();
				long elapsedTime = receiveTime - sendTime;
				String response = new String(receivePacket.getData(), 
	                                         0,
	                                         receivePacket.getLength());
				receiveCount++;
				
				long elapsedMilliseconds = elapsedTime / 1000000;
				times.add(elapsedMilliseconds);
				
				System.out.println("PING received from " + host + ": seq#=" + seq + " time=" + elapsedMilliseconds);
			} catch (Exception ex) {
				// The client did not receive a packet in the specified timeout window	
				System.out.println(ex.getMessage());
			}
		}
		
		// Gather and display statistics
		int dropCount = totalSent - receiveCount;
		
		System.out.println("--- PING statistics ---");
		System.out.println(totalSent + " packets transmitted, " + receiveCount + " received, " + (100 * dropCount / (double)totalSent) + "% packet loss");
		
		long min;
		long max;
		double avg;
		
		if (times.size() > 0) {
			min = times.getFirst();
			max = times.getFirst();
			
			double sum = 0;
			for (long time : times) {
				if (time > max) max = time;
				if (time < min) min = time;
				sum += time;
			}
			avg = sum / (double)times.size();

			System.out.println("rtt min/avg/max = " + min + " " + String.format("%.2f", avg) + " " + max + " ms");
		} else {
			System.out.println("Statistics unavailable -- no packets received.");
		}
		
		clientSocket.close();
	}
}